lib/galaxy/managers/credentials.py

```diff
@@ -0,0 +1,178 @@
+        )
+
+        return variables, secrets
+
+    def add_user_credentials(
+        self,
+        existing_user_credentials: List[Tuple[UserCredentials, CredentialsGroup]],
+        user_id: DecodedDatabaseIdField,
+        reference: str,
+        source_type: SOURCE_TYPE,
+        source_id: str,
+    ) -> DecodedDatabaseIdField:
+        user_credentials = next((uc[0] for uc in existing_user_credentials if uc[0].reference == reference), None)
+        if not user_credentials:
+            user_credentials = UserCredentials(
+                user_id=user_id,
+                reference=reference,
+                source_type=source_type,
+                source_id=source_id,
+            )
+            self.session.add(user_credentials)
+            self.session.flush()
+        return user_credentials.id
+
+    def add_group(
+        self,
+        existing_user_credentials: List[Tuple[UserCredentials, CredentialsGroup]],
+        user_credentials_id: DecodedDatabaseIdField,
+        group_name: str,
+        reference: str,
+    ) -> DecodedDatabaseIdField:
+        credentials_group = next(
+            (uc[1] for uc in existing_user_credentials if uc[1].name == group_name and uc[0].reference == reference),
+            None,
+        )
+        if not credentials_group:
+            credentials_group = CredentialsGroup(name=group_name, user_credentials_id=user_credentials_id)
+            self.session.add(credentials_group)
+            self.session.flush()
+        return credentials_group.id
+
+    def add_variable(
+        self,
+        variables: List[Variable],
+        user_credential_group_id: DecodedDatabaseIdField,
+        variable_name: str,
+        variable_value: str,
+    ) -> None:
+        variable = next(
+            (var for var in variables if var.name == variable_name),
+            None,
+        )
+        if variable:
+            variable.value = variable_value
+        else:
+            variable = Variable(
+                user_credential_group_id=user_credential_group_id,
+                name=variable_name,
+                value=variable_value,
+            )
+        self.session.add(variable)
+
+    def add_secret(
+        self,
+        secrets: List[Secret],
+        user_credential_group_id: DecodedDatabaseIdField,
+        secret_name: str,
+        secret_value: str,
+    ) -> None:
+        secret = next(
+            (sec for sec in secrets if sec.name == secret_name),
+            None,
+        )
+        if secret:
+            secret.already_set = True if secret_value else False
+        else:
+            secret = Secret(
+                user_credential_group_id=user_credential_group_id,
+                name=secret_name,
+                already_set=True if secret_value else False,
+            )
+        self.session.add(secret)
+
+    def commit_session(self) -> None:
```

Comment:
Also, I wouldn't expose a commit_session method on the manager: ideally, `session.commit()` should be only called from within the same context where the current transaction was started, so, in this case, from the credentials service module. I fear that having a manager method that commits the current session may lead to tangled transactions down the road. Also, it helps during debugging when we can very easily grep for when a session is committed - so `session.commit` is the safest option here, I think.